tests/test_jsvalidation.py

```diff
@@ -0,0 +1,597 @@
+    v1 = {
+        "$schema": "http://json-schema.org/draft-04/schema#",
+
+        "definitions": {
+            "address": {
+                "type": "object",
+                "properties": {
+                    "street_address": {"type": "string"},
+                    "city": {"type": "string"},
+                    "state": {"type": "string"}
+                },
+                "required": ["street_address", "city", "state"]
+            }
+        },
+
+        "type": "object",
+
+        "properties": {
+            "billing_address": {"$ref": "#/definitions/address"},
+            "shipping_address": {
+                "allOf": [
+                    {"$ref": "#/definitions/address"},
+                    {
+                        "properties":   {
+                            "type": {"enum": ["residential", "business"]}
+                        },
+                        "required": ["type"]
+                    }
+                ]
+            }
+        }
+    }
+    obj = JSONSchemaValidator()
+    obj.validate(v1, 'first')
+
+
+def test_ref_no_conflict_inside_schema_pass():
+    """Test that having no conflict in schema with references passes."""
+    v1 = {
+        "$schema": "http://json-schema.org/draft-04/schema#",
+
+        "definitions": {
+            "address": {
+                "type": "object",
+                "properties": {
+                    "street_address": {"type": "string"},
+                    "city": {"type": "string"},
+                    "state": {"type": "string"}
+                },
+                "required": ["street_address", "city", "state"]
+            }
+        },
+
+        "type": "object",
+
+        "properties": {
+            "billing_address": {"$ref": "#/definitions/address"},
+            "shipping_address": {
+                "allOf": [
+                    {"$ref": "#/definitions/address"},
+                    {
+                        "properties":   {
+                            "type": {"enum": ["residential", "business"]}
+                        },
+                        "required": ["type"]
+                    }
+                ]
+            }
+        }
+    }
+    obj = JSONSchemaValidator()
+    obj.validate(v1, 'first')
+
+
+def test_ref_conflict_inside_schema():
+    """Test that having a conflict in schema with references fails."""
+    v1 = {
+        "$schema": "http://json-schema.org/draft-04/schema#",
+
+        "definitions": {
+            "address": {
+                "type": "object",
+                "properties": {
+                    "street_address": {"type": "string"},
+                    "city": {"type": "string"},
+                    "state": {"type": "string"}
+                },
+                "required": ["street_address", "city", "state"]
+            }
+        },
+
+        "type": "object",
+
+        "properties": {
+            "billing_address": {"$ref": "#/definitions/differentaddress"},
```

Comment:
This reference is invalid. For this test to check compatibility, the JSON Schema itself should be valid.